presto-main-base/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarFromAnnotationsParser.java

```diff
@@ -155,4 +157,21 @@ public Set<Method> getMethods()
             return methods;
         }
     }
+
+    private static void checkPushdownSubfieldArgIndex(ParametricScalarImplementation implementation, ScalarImplementationHeader header, Method method)
+    {
+        Optional<Integer> pushdownSubfieldArgIndex = header.getHeader().getPushdownSubfieldArgIndex();
+        if (pushdownSubfieldArgIndex.isPresent()) {
+            Map<String, String> typeConstraintMapping = new HashMap<>();
+            Signature signature = implementation.getSignature();
+            for (TypeVariableConstraint constraint : signature.getTypeVariableConstraints()) {
+                typeConstraintMapping.put(constraint.getName(), constraint.getVariadicBound());
+            }
+
+            checkCondition(pushdownSubfieldArgIndex.get() >= 0, FUNCTION_IMPLEMENTATION_ERROR, "Method [%s] has negative pushdown subfield arg index", method);
+            checkCondition(signature.getArgumentTypes().size() > pushdownSubfieldArgIndex.get(), FUNCTION_IMPLEMENTATION_ERROR, "Method [%s] has out of range pushdown subfield arg index", method);
+            String typeVariableName = signature.getArgumentTypes().get(pushdownSubfieldArgIndex.get()).toString();
+            checkCondition(typeVariableName.equals(com.facebook.presto.common.type.StandardTypes.ROW) || typeConstraintMapping.get(typeVariableName).equals(com.facebook.presto.common.type.StandardTypes.ROW), FUNCTION_IMPLEMENTATION_ERROR, "Method [%s] does not have a struct or row type as pushdown subfield arg", method);
```

Comment:
same comment about maybe this is too restrictive and need to add some other conditions.